### ISSUE TITLE
The distance between features about Face Recognition 

### ISSUE THREAD

**Issue by jimeffry** [deepinsight/insightface#306]
First, thankyou for sharing the code. I have trained the project, and get the accuracy of 95% at the training dataset-ms1m. In order to achieve the goal of Face Recognition, I build a database of 20 peoples.  However, I get the number of distance is 0 always, when I give any one image to obtain the nearest person.
Next, I will show the training log, and the test log. In the test log, I will show the top 2 distances.
I'm looking forward to someone to answer， thankyou.

1. 

**INFO:root:Epoch[1] Batch [12000]	Speed: 312.45 samples/sec	acc=0.955898	lossvalue=0.000000
INFO:root:Epoch[1] Batch [13000]	Speed: 312.53 samples/sec	acc=0.955641	lossvalue=0.000001
INFO:root:Epoch[1] Batch [14000]	Speed: 312.49 samples/sec	acc=0.955789	lossvalue=0.000000
INFO:root:Epoch[1] Batch [15000]	Speed: 312.53 samples/sec	acc=0.956078	lossvalue=0.000000
INFO:root:Epoch[1] Batch [16000]	Speed: 312.51 samples/sec	acc=0.956133	lossvalue=0.000000
INFO:root:Epoch[1] Batch [17000]	Speed: 312.48 samples/sec	acc=0.955438	lossvalue=0.000000**

2. 
**deal with 43393/43399('distance ', [0.0, 0.0])
 deal with 43394/43399('distance ', [0.0, 0.0])
deal with 43395/43399('distance ', [0.0, 0.0])
deal with 43396/43399('distance ', [0.0, 0.0])
deal with 43397/43399('distance ', [0.0, 0.0])
deal with 43398/43399('distance ', [0.0, 0.0])
deal with 43399/43399('distance ', [0.0, 0.0])**

**Comment by Lxyzgf** [comment]
CUDA_VISIBLE_DEVICES='0,1' python -u train_softmax.py --data-dir $DATA_DIR --network "$NETWORK" --loss-type 5 --end-epoch 500000 --margin-a 0.9 --margin-m 0.4 --margin-b 0.15 --prefix "$PREFIX" --pretrained "$PREFIX",4 --display 1000 --verbose 10000  --lr 0.01  --ckpt 2 --per-batch-size 64 > "$LOGFILE"

**Comment by guoxf3z** [comment]
What's your distance metrics, how could you calculate distance

**Comment by Lxyzgf** [comment]
@guoxf3z 
I use the module annoy in python to build the database, and the type distance is 'angular'.

**Comment by Lxyzgf** [comment]
In my opinion， maybe the BatchNorm layer lead to the poor results. So next , I will try to remove the BatchNorm layer. Thankyou.

**Comment by guoxf3z** [comment]
I guess no mater how badly the model is trained, the distance should not be
0. The most probable reason of distance 0 might be using wrong output from
the network or wrong distance calculation formula.
Comparing a big action such as removing BN, it is worth to test something
small first, which is extracting whatever you used to calculate the
distance (are you really use embedding to calculate the distance?), observe
what is in them, and calculate distance manually using one-line cos
distance numpy method.
Best wishes.

**Comment by jimeffry** [comment]
Amazing， the features extraced from different person images, almost the same. Does anyone use this model for face recognition?

**Comment by nttstar** [comment]
@jimeffry This is unlikely to happen. Check your input setting.

**Comment by jimeffry** [comment]
@nttstar @guoxf3z The link is my test code. Please have a try, when you are convience. Thankyou!
[https://github.com/jimeffry/face-detect-recognize](url)

**Comment by jimeffry** [comment]
Here is the model. [https://pan.baidu.com/s/1appyFkJxpUQk_4iSWQQmZw](url)

**Comment by nttstar** [comment]
print img1 and img2 to check the value.

**Comment by antoine-cha** [comment]
Even an untrained network has low chance to give the same features for 2 different images, so there is most likely an error in your code which assigns the same features to different images.
To check for this kind of error
1. Pick 2 different images img1 and img2
2. Extract features for both using the same function
3. Compare the features. `d_l1 = np.sum(np.abs(feat1-feat2) )` will give you enough information.

If `d_l1 = 0.0` indeed you have an issue with your code. 
I glanced at your code, this is wrong: https://github.com/jimeffry/face-detect-recognize/blob/master/face_test.py#L65-L66

Also, you should not do that https://github.com/jimeffry/face-detect-recognize/blob/master/face_test.py#L63-L70
but 
```Python
def read_img(path_img):
    img = cv2.imread(path_img)
    img = cv2.cvtColor(img,cv2.COLOR_BGR2RGB)
    img = (img-127.5)*0.0078125
    return img

img1 = read_img(path1)
img2 = read_img(path2)
```
It is less error prone since all the images should go through the same processing pipeline.

Hope this can help, happy debugging !

**Comment by jimeffry** [comment]
@antoine-cha 
I am grateful for your reply. I had run the same images and different images before, and printed the features. According to the results, I found that the results almost the same.

1. 
test, img1 and img1, below is the part of the features
**('feat1 ', array([-0.00628504, 0.00307494, 0.00274607, -0.00111485, 0.00222094],dtype=float32))
  ('feat2 ', array([-0.00628504, 0.00307494, 0.00274607, -0.00111485,0.00222094],dtype=float32))**

2. 
test, img1 and img2, the part result is
**('feat1 ', array([-0.00628504, 0.00307494, 0.00274607, -0.00111485, 0.00222094],dtype=float32))
 ('feat2 ', array([-0.00628524, 0.00307335, 0.00274694, -0.00111421, 0.00222139],dtype=float32))**

**Comment by nttstar** [comment]
do not subtract mean by using insightface models.

**Comment by jimeffry** [comment]
@antoine-cha 
Could you share it, if you already have trained a good model. Thank  you very much.
I have trained a model using the sphereface in caffe. In terms with the testing result, I am pleased .

**Comment by jimeffry** [comment]
@nttstar 
I really appreciate your work! I will keep on trying.

**Comment by Lucien7786** [comment]
@jimeffry you have done twice norm operation on your input data
    def extractfeature(self,img):
        h_,w_,chal_ = img.shape
        if h_ !=self.h or w_ !=self.w:
            img = cv2.resize(img,(self.w,self.h))
        img = (img-127.5)*0.0078125
        img = np.transpose(img,(2,0,1))
        img = np.expand_dims(img,0)

def read_img(path_img):
    img = cv2.imread(path_img)
    img_org = img
    img = cv2.cvtColor(img,cv2.COLOR_BGR2RGB)
    img = (img-127.5)*0.0078125
return img,img_org

-----------------

def compare_img(path1,path2,model_path):
    parm = args()
img1,img_1 = read_img(path1)
...
    if img1 is not None and img2 is not None:
feature_1 = FaceModel.extractfeature(img1)

**Comment by jimeffry** [comment]
@Lucien7786 
I really appreciate your help. This question has plagued me for a long time. 
Finally， I get the point  that the method of image processing is consistent before and after training.

**Comment by barzan-hayati** [comment]
> ```python
> img = cv2.cvtColor(img,cv2.COLOR_BGR2RGB)
> ```


> img = cv2.cvtColor(img,cv2.COLOR_BGR2RGB)

Is it necessary to change color image before doing extractfeature?




> feature_1 = FaceModel.extractfeature(img1)
